src/mono/mono/eventpipe/ep-rt-mono.c

```diff
@@ -2857,6 +2857,433 @@ ep_rt_mono_write_event_ee_startup_start (void)
+}
+
+// ETW has a limitation of 64K for TOTAL event Size, however there is overhead associated with
+// the event headers.   It is unclear exactly how much that is, but 1K should be sufficiently
+// far away to avoid problems without sacrificing the perf of bulk processing.
+static const uint32_t cb_max_etw_event = 63 * 1024;
+
+// Estimate of how many bytes we can squeeze in the event data for the value struct
+// array.  (Intentionally overestimate the size of the non-array parts to keep it safe.)
+static const uint32_t k_max_bytes_type_values = (cb_max_etw_event - 0x30);
+
+// Estimate of how many type value elements we can put into the struct array, while
+// staying under the ETW event size limit. Note that this is impossible to calculate
+// perfectly, since each element of the struct array has variable size.
+//
+// In addition to the byte-size limit per event, Windows always forces on us a
+// max-number-of-descriptors per event, which in the case of BulkType, will kick in
+// far sooner. There's a max number of 128 descriptors allowed per event. 2 are used
+// for Count + ClrInstanceID. Then 4 per batched value. (Might actually be 3 if there
+// are no type parameters to log, but let's overestimate at 4 per value).
+static const uint32_t k_max_count_type_values = (128 - 2) / 4;
+
+struct _BulkTypeEventLogger {
+	BulkTypeValue m_rg_bulk_type_values[k_max_count_type_values];
+	uint32_t m_n_bulk_type_value_count;
+	uint32_t m_n_bulk_type_value_byte_count;
+	uint8_t *m_p_bulk_type_event_buffer[65536];
+};
+
+//---------------------------------------------------------------------------------------
+//
+// ep_rt_mono_fire_bulk_type_event fires an ETW event for all the types batched so far,
+// it then resets the state to start batching new types at the beginning of the
+// m_rg_bulk_type_values array.
+//
+
+void
+ep_rt_mono_fire_bulk_type_event (BulkTypeEventLogger *p_type_logger)
+{
+	if (p_type_logger->m_n_bulk_type_value_count == 0)
+		return;
+	memset(p_type_logger->m_p_bulk_type_event_buffer, 0, 65536 * sizeof(uint8_t));
+	uint16_t n_clr_instance_id = clr_instance_get_id();
+
+	uint32_t i_size = 0;
+
+	for (int i_type_data = 0; i_type_data < p_type_logger->m_n_bulk_type_value_count; i_type_data++)
+	{
+		BulkTypeValue *target = &p_type_logger->m_rg_bulk_type_values[i_type_data];
+
+		memcpy(p_type_logger->m_p_bulk_type_event_buffer + i_size,
+			   &target->fixed_sized_data,
+			   sizeof(target->fixed_sized_data));
+		i_size += sizeof(target->fixed_sized_data);
+
+		char *wsz_name = target->s_name;
+		if (!wsz_name || strlen(wsz_name) == 0)
+		{
+			p_type_logger->m_p_bulk_type_event_buffer[i_size++] = 0;
+			p_type_logger->m_p_bulk_type_event_buffer[i_size++] = 0;
+		}
+		else
+		{
+			uint32_t name_size = (strlen(target->s_name) + 1) * sizeof(wchar_t);
+			memcpy(p_type_logger->m_p_bulk_type_event_buffer + i_size, wsz_name, name_size);
+			i_size += strlen(wsz_name);
```

Comment:
Ah, the `wsz_name` is a typo, it should use `name_size`.
Things mainly weren't breaking since s_name isn't ever set. I haven't fully understood the snippet in https://github.com/dotnet/runtime/pull/68571#discussion_r863343068 to know if we need to do the same on mono yet.

Reply:
Changed to name_size